### PR TITLE
[REVIEW] CSV Reader: Add parameter to specify the byte range of the input file to the read

### LINKED ISSUE #568: [FEA] Support byte ranges in read_csv (and other formats)
To enable batched reading of large text files (csv, line-delimited-json, ...) it is often useful to be able to read a chunk of data while respecting some delimiter.  For example some process might be tasked with reading from 4GB to 5GB in the file starting and stopping at a `\n` character.

This would be particularly helpful when parallelizing around `cudf.read_csv` with `dask_cudf.read_csv`, but would presumably also be helpful for other functions in the future.  

The dask.bytes module does this here:

https://github.com/dask/dask/blob/master/dask/bytes/utils.py#L149-L238

However it returns Python `bytes` objects so might not be appropriate for `cudf`.  It might be good to take inspiration from though.

### PR DESCRIPTION
**Pull request by vuule**
Added a list-like parameter to read_csv() and  read_csv_strings() to allow user to specify the byte range within the  input file to be read. This allows users to split the CSV parsing into  multiple jobs and dispatch independently. 
* First element in the list specifies the byte offset, the second one specifies the size of the range. 
* Currently the feature is not compatible with skiprows, nrows,  skipfooter because it does not make sense to mix the row and byte count  when reading a CSV file. 
* At the end of the range, any row starting before or at the end will be included in the range. 
* The row that starts on or before the start of the range will not be included in the range, except when the range offset is zero. 

Fixes #568

### CODE REVIEW

**Comment by harrism** [comment]
rerun tests

File: cpp/src/io/csv/csv_reader.cu [hunk 1]
```diff
@@ -785,6 +843,38 @@ gdf_dtype convertStringToDtype(std::string &dtype) {
...
+	// TODO: don't copy the whole array, only up to calculateMaxRowSize() rows are relevant here
+	vector<cu_recstart_t> h_rec_starts(raw_csv->num_records + 1);
+	const size_t rec_start_size = sizeof(cu_recstart_t) * (h_rec_starts.size());
+	CUDA_TRY( cudaMemcpy(h_rec_starts.data(), raw_csv->recStart, rec_start_size, cudaMemcpyDeviceToHost) );
+	size_t rec_start_idx = raw_csv->num_records;
+
+	// TODO: we could use a modified binary search here instead
+	// Searching for the last row that starts before the range end
+	// This row with be the extra one we want to read, as it ends after the byte range end
+	while (rec_start_idx != 0 && h_rec_starts[rec_start_idx] > (cu_recstart_t)byte_range_size) {
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/csv/csv_reader.cu]: @harrism Really depends on the number of columns and the data size in each cell. 
Some estimates:
* 2 columns -> ~400 or less;
* 50 columns -> ~20 or less;
* 1000 columns -> ~6 or less;
I can use a binary search here, but it's very unlikely to provide significant benefit.

File: cpp/src/io/csv/csv_reader.cu [hunk 2]
```diff
@@ -134,6 +150,7 @@ using string_pair = std::pair<const char*,size_t>;
 //
 gdf_error parseArguments(csv_read_arg *args, raw_csv_t *csv);
 // gdf_error getColNamesAndTypes(const char **col_names, const  char **dtypes, raw_csv_t *d);
+gdf_error trimExtraRows(gdf_size_type byte_range_size, raw_csv_t * raw_csv);
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/csv/csv_reader.cu]: I was just following the existing pattern in the document. I will gladly remove the declaration.

**Comment by vuule** [comment]
@harrism Applied your suggestions, please review.

**Comment by harrism** [comment]
@vuule There is a simple conflict but I'm not sure which is the correct resolution so please take care of it and then ping me.

**Pull request merged**